### DISCUSSION TITLE
Improve $NOW adjustments

### DISCUSSION THREAD

**Discussion by ijpatricio** [directus/directus#11314] [Archive]
As of today, the possible adjustments are

Adding or subtracting X amount of units of time.

Example: ($NOW is 2021-12-24 19:00:00) 
```
$NOW(+1 day) => 2021-12-25 19:00:00

$NOW(+1 month) => 2022-01-24 19:00:00

$NOW(+1day) => 2022-12-24 19:00:00
```

And that's just fine!

What I think could be improved is to add relative timeframes.
Say in `validation`, if you want a certain date to be:
- greater than tomorrow... +1 day starts 24h after $now
- on the following year... +1 year starts 365 days and 24h after $now

Without a `hook` (extension) this seems not possible to do.

Having in mind to boost the `no code` capability, relative timeframes would be amazing, right?

Imagine this:

Example: ($NOW is 2021-12-24 19:00:00) 
```
$NOW(tomorrow) => 2021-12-25 00:00:00

$NOW(3 months from now) => 2022-03-24 00:00:00

$NOW(next month) => 2022-01-01 00:00:00

$NOW(next year) => 2023-01-01 00:00:00
```

Is this something that sounds interesting to you??

Thanks in advance! 😍

**Comment by ijpatricio** [comment]
On the technical side of things, `date-fns` is under the hood.

I'm not sure if that fluent kind of phrasing is easily feasible, but I believe a `FORMAT` is somewhat easy to achieve.

Bear with me 😄 

Achieving "Tomorrow"
```
before:
$NOW(+ 1 day)

after:
$NOW(+ 1 day | format: YYYY-MM-DD 00:00:00 )
```


Achieving "Next month"
```
before:
$NOW(+ 1 month)

after:
$NOW(+ 1 month | format: YYYY-MM-01 00:00:00 )
```

It's a little less fluent and not as expressive, but arguably it's a good way to go, as it is simple to implement.

Any thoughts??

**Reply by benhaynes** [reply]
Just to be clear, the same functionality is possible in both... correct? This is just an alternate syntax? I defer to @rijkvanzanten on this one.

**Reply by ijpatricio** [reply]
yep, it's an additive thing. use "as is", or also apply a "format", just the way Vue does with filters, using a pipe `|`

**Comment by rijkvanzanten** [comment]
> Say in validation, if you want a certain date to be:
> 
> greater than tomorrow... +1 day starts 24h after $now
> on the following year... +1 year starts 365 days and 24h after $now
> Without a hook (extension) this seems not possible to do.

This should already be possible, using the same $NOW logic. The filters used in the API for regular operations rely on the same logic as the validation engine, so that should be all the same 🙂

**Reply by ijpatricio** [reply]
Hey Rijk! It's a long time we didn't speak! I'm so sorry, but I could not understand your point.

Yes, I know today we can do $NOW(+1 day). I've it, and I could not achieve the business rule because `+1 day` adds 24h exactly.

When creating the record, the business rule is that the `start_date` field must be at least on the following day (which starts at 00h00).
If you fill the form `today` at 16h00 and put "tomorrow date" in the date picker, validation will fail because that will set it at 00h00.

Do you see now what I'm trying to accomplish?

My 2nd post is a suggestion of how we could achieve that, with a bit of change to in [adjust-date-ts](https://github.com/directus/directus/blob/main/packages/shared/src/utils/adjust-date.ts)

I would be happy to do a PR.

No breaking changes, it would just apply a format to the resulting date, delegating to the native `date-fns` format thing.

**Reply by ijpatricio** [reply]
If I understood wrong, please tell me how to set a validation field for "greater than or equal to tomorrow at 00h00."

**Reply by rijkvanzanten** [reply]
Ah yeah, the "at 00:00" part is not currently possible 👍🏻

**Reply by rijkvanzanten** [reply]
Almost wondering if that should be it's own function, like `setTime($NOW, 00:00)` 🤔

**Reply by jamescammarano** [reply]
Would we be able to add that like we did the date functions?

**Reply by ijpatricio** [reply]
yesss!!! now we're on the same page!

I think there's a better chance with `format`...

because `00:00` would "reset" the time....

but what about when you want "next month at day 1 at 00:00" ??

that's where 

```
$NOW(+ 1 month | format: YYYY-MM-01 00:00:00 )
```

Notice that the format that I wrote is not the usual "YYYY-MM-DD H:i:s" , but I override with hardcoded numbers on the day and time

`Format` is for sure in `date-fns`, the library underneath.

If you like the idea, I'll definitely give it a go for a PR, after I finish this "Directus + XXXXX" setup.

What do you say?

**Reply by ijpatricio** [reply]
> Would we be able to add that like we did the date functions?

@jaycammarano  If I understood correctly, yes!

Let me clarify what I understood: we go to the raw field and add something like this

![image](https://user-images.githubusercontent.com/26031459/151664697-a972dc7d-e425-4d97-abe0-59c6a80ce8ed.png)

Is this what you were asking?

PS: this modal to edit raw value could also be bigger, and a JSON editor/linter field, right??

**Reply by northern-mindset** [reply]
> This should already be possible, using the same $NOW logic. The filters used in the API for regular operations rely on the same logic as the validation engine

Does this mean that we can use `$NOW` in the field validation filters? Doesn't seem so, I'm getting `[INTERNAL_SERVER_ERROR] date must have a valid date format or reference` when trying to save the item with `$NOW` in validation filters. What's wrong?

![image](https://github.com/directus/directus/assets/6065071/b20fef0e-c482-4144-a5c8-c097cb33e647)

![image](https://github.com/directus/directus/assets/6065071/3d040caf-3f57-47e9-906a-f2b41cf992e4)

In the Content Module filtering works with `$NOW`.

Directus v10.2.1 (self-hosted via Docker)

**Comment by licitdev** [comment]
Posting this as a reference to a past thread about an improvement to `$NOW()` where users could define the start or end of a "current" time period. https://github.com/directus/directus/issues/8529

Examples:
**current month**
01/01/2022 00:00 to 31/01/2022 23:59 with `$NOW(0, start month)` and `$NOW(0, end month)`

**next year**
01/01/2023 00:00 to 31/12/2023 23:59 with `$NOW(+1, start year)` and `$NOW(+1, end year)`

**previous day**
30/01/2022 00:00 to 31/01/2022 23:59 with `$NOW(-1, start day)` and `$NOW(-1, end day)`

**Comment by ijpatricio** [comment]
Hey @licitdev 

I reference this discussion on the PR, but it didn't appear here. It's here https://github.com/directus/directus/pull/11340

I think you will get better insights from the PR, and associated changes to the docs.
It also has a  reference to what we can do with the `format` function.

I read the #8529, but I didn't get an actionable thing I could do to help further.

Is there anything I can help with?

**Reply by licitdev** [reply]
Hello @ijpatricio, as this is a discussion on the `$NOW` topic, I thought of linking it here for reference. Your PR #11340 allows for formatting of $NOW with adjustments, which opens up new possibilities for use. 🙌🏼 

I was reminded of that past conversation, about a new functionality where the `start` and `end` of time periods can be used in filters. Your current PR would allow for the retrieval of `start` and `end` date for both days and years. 👍🏻 

However, I wonder how we could access the end of the month through this method, as there are differing number of days in different months.

**Reply by ijpatricio** [reply]
Hey @licitdev , I'm so glad I could help!

Well, not sure if I get it out of the bat, would have to refer to the `data-fns` library, but before that... isn't the end of the current month, the moment before the next of the following??

E.g. if you want `before the end of current month`, isn't it:  `_lt: $NOW(+ 1 month | format: YYYY-MM-01 00:00:00 )` ?

**Reply by licitdev** [reply]
@ijpatricio Yes you're right!! I was thinking of `_lte`. Thank you!! 👍🏻

**Reply by ijpatricio** [reply]
You're very welcome! 👍

**Comment by iamdagy** [comment]
This would be a very useful feature to have. The format option is flexible I think it could also generate date from $NOW without the time ->  $NOW(  format: YYYY-MM-DD ).
